packages/ses/src/lockdown-shim.js

```diff
@@ -90,52 +92,69 @@ export function lockdown(options = {}) {
-  defineProperties(globalThis, tameGlobalErrorObject(errorTaming).start);
-  defineProperties(globalThis, tameGlobalMathObject(mathTaming).start);
-  defineProperties(globalThis, tameGlobalRegExpObject(regExpTaming).start);
+  intrinsicsMgr.addIntrinsics(tameFunctionConstructors());
 
-  /**
-   * 2. WHITELIST to standardize the environment.
-   */
+  intrinsicsMgr.addIntrinsics(tameDateConstructor(dateTaming));
+  intrinsicsMgr.addIntrinsics(tameErrorConstructor(errorTaming));
+  intrinsicsMgr.addIntrinsics(tameMathObject(mathTaming));
+  intrinsicsMgr.addIntrinsics(tameRegExpConstructor(regExpTaming));
+
+  intrinsicsMgr.addIntrinsics(getAnonymousIntrinsics());
 
-  // Extract the intrinsics from the global.
-  const intrinsics = getIntrinsics();
+  intrinsicsMgr.completePrototypes();
+
+  const intrinsics = intrinsicsMgr.finalIntrinsics();
 
   // Replace *Locale* methods with their non-locale equivalents
   tameLocaleMethods(intrinsics, localeTaming);
 
+  /**
+   * 2. WHITELIST to standardize the environment.
+   */
+
   // Remove non-standard properties.
   whitelistIntrinsics(intrinsics);
 
   // Repair problems with legacy accessors if necessary.
   repairLegacyAccessors();
 
+  // Initialize the powerful initial global, i.e., the global of the
+  // start compartment, from the intrinsics.
+  initGlobalObject(globalThis, intrinsics, initialGlobalPropertyNames, {});
+
   /**
    * 3. HARDEN to share the intrinsics.
    */
 
-  // Circumvent the override mistake.
-  const detachedProperties = enablePropertyOverrides(intrinsics);
+  function hardenIntrinsics() {
+    // Circumvent the override mistake.
+    const detachedProperties = enablePropertyOverrides(intrinsics);
 
-  // Finally register and optionally freeze all the intrinsics. This
-  // must be the operation that modifies the intrinsics.
-  lockdownHarden(intrinsics);
-  lockdownHarden(detachedProperties);
+    // Finally register and optionally freeze all the intrinsics. This
+    // must be the operation that modifies the intrinsics.
+    lockdownHarden(intrinsics);
+    lockdownHarden(detachedProperties);
 
-  // Having completed lockdown without failing, the user may now
-  // call `harden` and expect the object's transitively accessible properties
-  // to be frozen out to the fringe.
-  // Raise the `harden` gate.
-  lockedDown = true;
+    // Having completed lockdown without failing, the user may now
+    // call `harden` and expect the object's transitively accessible properties
+    // to be frozen out to the fringe.
+    // Raise the `harden` gate.
+    lockedDown = true;
 
-  // Returning `true` indicates that this is a JS to SES transition.
-  return true;
+    // Returning `true` indicates that this is a JS to SES transition.
+    return true;
+  }
+  return hardenIntrinsics;
 }
+
+export const lockdown = (options = {}) => {
+  const optHardenIntrinsics = repairIntrinsics(options);
+  if (optHardenIntrinsics === false) {
+    return false;
+  }
```

Comment:
This will probably be easier to force through type discipline if it’s “optional function” rather than “function or bool (and only ever specifically false)”.

This seems odd. I’ll try to find the rationale. Weird enough for a comment.

Reply:
Posting a commit to this that keeps the two-stage lockdown without branching or type unions. `repairIntrinsics` returns a no-op (returns false) function called `alreadyHardenedIntrinsics` that serves both as an indicator and a callable function that satisfies the same type as `hardenIntrinsics`. `repairIntrinsics` returns a function named `maybeHardenIntrinsics` that is either `hardenIntrinsics` or `alreadyHardenedIntrinsics`.